lib/internal/fs/watch/linux.js

```diff
@@ -0,0 +1,259 @@
+'use strict';
+
+const {
+  SafeMap,
+  StringPrototypeStartsWith,
+  SymbolAsyncIterator,
+} = primordials;
+
+const {
+  codes: {
+    ERR_FEATURE_UNAVAILABLE_ON_PLATFORM,
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { getValidatedPath } = require('internal/fs/utils');
+const { kFSWatchStart } = require('internal/fs/watchers');
+const { kEmptyObject } = require('internal/util');
+const { validateObject, validateBoolean, validateAbortSignal } = require('internal/validators');
+const { EventEmitter, once } = require('events');
+const path = require('path');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = new SafeMap()) {
+  const { opendir } = lazyLoadFsPromises();
+
+  const filenames = await opendir(dir);
+
+  for await (const file of filenames) {
+    const f = path.join(dir, file.name);
+
+    files.set(f, file);
+
+    if (file.isDirectory()) {
+      await traverse(f, files);
+    }
+  }
+
+  return files;
+}
+
+class FSWatcher extends EventEmitter {
+  #options = null;
+  #closed = false;
+  #files = new SafeMap();
+  #rootPath = path.resolve();
+
+  constructor(options = kEmptyObject) {
+    super();
+
+    validateObject(options, 'options');
+
+    if (options.persistent != null) {
+      validateBoolean(options.persistent, 'options.persistent');
+    }
+
+    if (options.recursive != null) {
+      validateBoolean(options.recursive, 'options.recursive');
+    }
+
+    if (options.signal != null) {
+      validateAbortSignal(options.signal, 'options.signal');
+    }
+
+    if (options.encoding != null) {
+      // This is required since on macOS and Windows it throws ERR_INVALID_ARG_VALUE
+      if (typeof options.encoding !== 'string') {
+        throw new ERR_INVALID_ARG_VALUE('encoding', 'options.encoding');
+      }
+    }
+
+    const { persistent, recursive, signal, encoding } = options;
+    this.#options = { persistent, recursive, signal, encoding };
+  }
+
+  close() {
+    if (this.#closed) {
+      return;
+    }
+
+    const { unwatchFile } = lazyLoadFsSync();
+    this.#closed = true;
+
+    for (const file of this.#files.keys()) {
+      unwatchFile(file);
+    }
+
+    this.#files.clear();
+    this.emit('close');
+  }
+
+  /**
+   * @param {string} file
+   * @return {string}
+   */
+  #getPath(file) {
+    const root = this.#files.get(this.#rootPath);
+
+    // When watching files the path is always the root.
+    if (root.isFile()) {
+      return this.#rootPath.slice(this.#rootPath.lastIndexOf('/') + 1);
+    }
+
+    if (file === this.#rootPath) {
+      return this.#rootPath;
+    }
```

Comment:
Can we move the check higher to skip the previous checks?

Reply:
This would not work. Since, if watch recursive takes a file name, it always emits the file name, not the path.

Reply:
I don't really understand what this method is doing, could you update its description to add a sentence explaining what this is for?

Reply:
@aduh95 I added comments to `#getPath()` function